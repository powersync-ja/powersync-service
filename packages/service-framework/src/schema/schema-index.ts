export * from './definitions.js';

export * from './json-schema/keywords.js';
export * from './json-schema/parser.js';
export * from './validators/schema-validator.js';
export * from './validators/ts-codec-validator.js';
