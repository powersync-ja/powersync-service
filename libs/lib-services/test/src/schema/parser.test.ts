import { describe, test, expect } from 'vitest';

import * as framework_schema from '../../../src/schema/schema-index.js';

describe('schema-tools', () => {
  test('it should correctly prune unused definitions', () => {
    const schema = framework_schema.parseJSONSchema({
      definitions: {
        // unused, should be stripped out
        a: {
          type: 'object',
          properties: {
            prop: {
              type: 'string'
            }
          },
          required: ['prop']
        },

        // extended reference, should be included after walking the full schema
        b: {
          type: 'object',
          properties: {
            prop: {
              type: 'string'
            }
          },
          required: ['prop']
        },
        b1: {
          type: 'object',
          properties: {
            prop: {
              $ref: '#/definitions/b'
            }
          },
          required: ['prop']
        },

        // circular reference, should not result in the walker getting stuck
        c: {
          type: 'object',
          properties: {
            prop: {
              $ref: '#/definitions/c'
            }
          },
          required: ['prop']
        }
      },

      type: 'object',
      properties: {
        a: {
          type: 'object',
          properties: {
            a: {
              $ref: '#/definitions/b1'
            },
            b: {
              enum: ['A']
            }
          }
        },
        b: {
          oneOf: [
            {
              type: 'object',
              properties: {
                a: {
                  $ref: '#/definitions/c'
                }
              },
              required: ['a'],
              additionalProperties: false
            }
          ]
        }
      }
    });

    expect(schema.compile()).toMatchSnapshot();
  });
});
