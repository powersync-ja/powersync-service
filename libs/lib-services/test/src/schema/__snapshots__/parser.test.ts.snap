// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema-tools > it should correctly prune unused definitions 1`] = `
{
  "definitions": {
    "b": {
      "properties": {
        "prop": {
          "type": "string",
        },
      },
      "required": [
        "prop",
      ],
      "type": "object",
    },
    "b1": {
      "properties": {
        "prop": {
          "$ref": "#/definitions/b",
        },
      },
      "required": [
        "prop",
      ],
      "type": "object",
    },
    "c": {
      "properties": {
        "prop": {
          "$ref": "#/definitions/c",
        },
      },
      "required": [
        "prop",
      ],
      "type": "object",
    },
  },
  "properties": {
    "a": {
      "properties": {
        "a": {
          "$ref": "#/definitions/b1",
        },
        "b": {
          "enum": [
            "A",
          ],
        },
      },
      "type": "object",
    },
    "b": {
      "oneOf": [
        {
          "additionalProperties": false,
          "properties": {
            "a": {
              "$ref": "#/definitions/c",
            },
          },
          "required": [
            "a",
          ],
          "type": "object",
        },
      ],
    },
  },
  "type": "object",
}
`;
